File: python-sdk/agentscope/coding_agent.py
from __future__ import annotations

import difflib
import subprocess
import uuid
from dataclasses import dataclass
from functools import wraps
from pathlib import Path
from typing import Any, Callable, TypeVar

from .instrumentation import auto_instrument
from .run import _current_run_state, observe_run
from .span import observe_span

F = TypeVar("F", bound=Callable[..., Any])


@dataclass
class CommandResult:
    command: str
    exit_code: int
    stdout: str
    stderr: str


def _append_artifact(*, span: dict[str, Any], kind: str, payload: dict[str, Any]) -> None:
    run_state = _current_run_state()
    if run_state is None:
        return

    run_state.artifacts.append(
        {
            "id": str(uuid.uuid4()),
            "run_id": span["run_id"],
            "span_id": span["id"],
            "kind": kind,
            "payload": payload,
        }
    )


def _as_path(file_path: str | Path) -> Path:
    return file_path if isinstance(file_path, Path) else Path(file_path)


class coding_agent_run:
    def __init__(self, agent_name: str = "coding_agent") -> None:
        self.agent_name = agent_name
        self._run = observe_run("coding_agent", agent_name=agent_name)

    def __enter__(self) -> dict[str, Any]:
        auto_instrument()
        return self._run.__enter__()

    def __exit__(self, exc_type, exc, tb) -> bool:
        return self._run.__exit__(exc_type, exc, tb)


def instrument_coding_agent(fn: F) -> F:
    @wraps(fn)
    def _wrapped(*args: Any, **kwargs: Any) -> Any:
        with coding_agent_run(agent_name=fn.__name__):
            return fn(*args, **kwargs)

    return _wrapped  # type: ignore[return-value]


def read_file(file_path: str | Path, *, encoding: str = "utf-8") -> str:
    path = _as_path(file_path)
    with observe_span("file_read") as span:
        span["metadata"] = {"file_path": str(path)}
        return path.read_text(encoding=encoding)


def write_file(file_path: str | Path, content: str, *, encoding: str = "utf-8") -> None:
    path = _as_path(file_path)
    previous = path.read_text(encoding=encoding) if path.exists() else ""

    with observe_span("file_write") as span:
        span["metadata"] = {"file_path": str(path)}
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text(content, encoding=encoding)

        diff = "".join(
            difflib.unified_diff(
                previous.splitlines(keepends=True),
                content.splitlines(keepends=True),
                fromfile=f"a/{path}",
                tofile=f"b/{path}",
            )
        )
        _append_artifact(
            span=span,
            kind="file.diff",
            payload={"file_path": str(path), "diff": diff},
        )
        _append_artifact(
            span=span,
            kind="file.content",
            payload={"file_path": str(path), "content": content},
        )


def run_command(
    command: str | list[str],
    *,
    cwd: str | Path | None = None,
    env: dict[str, str] | None = None,
    check: bool = False,
    shell: bool | None = None,
) -> CommandResult:
    resolved_shell = shell if shell is not None else isinstance(command, str)

    with observe_span("command_exec") as span:
        command_text = command if isinstance(command, str) else " ".join(command)
        completed = subprocess.run(
            command,
            cwd=str(cwd) if cwd is not None else None,
            env=env,
            check=False,
            capture_output=True,
            text=True,
            shell=resolved_shell,
        )
        span["metadata"] = {
            "command": command_text,
            "exit_code": completed.returncode,
            "stdout": completed.stdout,
            "stderr": completed.stderr,
        }

        if check and completed.returncode != 0:
            raise subprocess.CalledProcessError(
                completed.returncode,
                command,
                output=completed.stdout,
                stderr=completed.stderr,
            )

        return CommandResult(
            command=command_text,
            exit_code=completed.returncode,
            stdout=completed.stdout,
            stderr=completed.stderr,
        )

File: ts-sdk/src/coding_agent.ts
import { exec, execFile } from "node:child_process";
import { promises as fs } from "node:fs";
import { dirname } from "node:path";
import { promisify } from "node:util";

import { autoInstrument } from "./instrumentation";
import { observeRun } from "./run";
import { addArtifact, observeSpan } from "./span";
import type { CommandResult, ObserveRunOptions } from "./types";

const execAsync = promisify(exec);
const execFileAsync = promisify(execFile);

export interface RunCommandOptions {
  cwd?: string;
  env?: NodeJS.ProcessEnv;
  check?: boolean;
  shell?: boolean;
}

export async function codingAgentRun<T>(
  fn: () => Promise<T> | T,
  options: Omit<ObserveRunOptions, "agentName"> & { agentName?: string } = {},
): Promise<T> {
  autoInstrument();
  return observeRun("coding_agent", fn, {
    ...options,
    agentName: options.agentName ?? "coding_agent",
  });
}

export function instrumentCodingAgent<TArgs extends unknown[], TResult>(
  fn: (...args: TArgs) => Promise<TResult> | TResult,
): (...args: TArgs) => Promise<TResult> {
  return async (...args: TArgs) => {
    return codingAgentRun(() => fn(...args), { agentName: fn.name || "coding_agent" });
  };
}

export async function readFile(filePath: string, encoding: BufferEncoding = "utf-8"): Promise<string> {
  return observeSpan("file_read", async () => {
    return fs.readFile(filePath, { encoding });
  }, {
    metadata: { file_path: filePath },
  });
}

export async function writeFile(
  filePath: string,
  content: string,
  encoding: BufferEncoding = "utf-8",
): Promise<void> {
  const previous = await fs.readFile(filePath, { encoding }).catch(() => "");

  await observeSpan("file_write", async () => {
    await fs.mkdir(dirname(filePath), { recursive: true });
    await fs.writeFile(filePath, content, { encoding });

    addArtifact("file.diff", {
      file_path: filePath,
      diff: createUnifiedDiff(filePath, previous, content),
    });
    addArtifact("file.content", {
      file_path: filePath,
      content,
    });
  }, {
    metadata: { file_path: filePath },
  });
}

export async function runCommand(
  command: string | string[],
  options: RunCommandOptions = {},
): Promise<CommandResult> {
  const resolvedShell = options.shell ?? typeof command === "string";
  const commandText = typeof command === "string" ? command : command.join(" ");
  if (Array.isArray(command) && command.length === 0) {
    throw new Error("runCommand requires at least one command part when command is an array");
  }

  return observeSpan("command_exec", async () => {
    try {
      const result = typeof command === "string" || resolvedShell
        ? await execAsync(commandText, {
          cwd: options.cwd,
          env: options.env,
          shell: resolvedShell ? (process.env.SHELL ?? "/bin/sh") : undefined,
        })
        : await execFileAsync(command[0], command.slice(1), {
          cwd: options.cwd,
          env: options.env,
          shell: false,
        });

      const stdout = toText(result.stdout);
      const stderr = toText(result.stderr);
      addArtifact("command.stdout", { command: commandText, stdout });
      if (stderr) {
        addArtifact("command.stderr", { command: commandText, stderr });
      }

      return {
        command: commandText,
        exitCode: 0,
        stdout,
        stderr,
      };
    } catch (error) {
      const stdout = readExecErrorField(error, "stdout");
      const stderr = readExecErrorField(error, "stderr");
      const code = readExecErrorCode(error);

      addArtifact("command.stdout", { command: commandText, stdout });
      if (stderr) {
        addArtifact("command.stderr", { command: commandText, stderr });
      }

      if (options.check) {
        throw error;
      }

      return {
        command: commandText,
        exitCode: code,
        stdout,
        stderr,
      };
    }
  }, {
    metadata: { command: commandText },
  });
}

function createUnifiedDiff(filePath: string, previous: string, next: string): string {
  if (previous === next) {
    return "";
  }

  const before = previous.split("\n");
  const after = next.split("\n");
  const lines: string[] = [`--- a/${filePath}`, `+++ b/${filePath}`, "@@"];

  const maxLen = Math.max(before.length, after.length);
  for (let i = 0; i < maxLen; i += 1) {
    const prevLine = before[i];
    const nextLine = after[i];

    if (prevLine === nextLine) {
      if (prevLine !== undefined) {
        lines.push(` ${prevLine}`);
      }
      continue;
    }

    if (prevLine !== undefined) {
      lines.push(`-${prevLine}`);
    }
    if (nextLine !== undefined) {
      lines.push(`+${nextLine}`);
    }
  }

  return `${lines.join("\n")}\n`;
}

function readExecErrorField(error: unknown, key: "stdout" | "stderr"): string {
  if (error && typeof error === "object" && key in error) {
    const value = (error as Record<string, unknown>)[key];
    return toText(value);
  }
  return "";
}

function readExecErrorCode(error: unknown): number {
  if (error && typeof error === "object" && "code" in error) {
    const value = (error as Record<string, unknown>).code;
    if (typeof value === "number" && Number.isFinite(value)) {
      return value;
    }
    if (typeof value === "string") {
      const parsed = Number.parseInt(value, 10);
      if (!Number.isNaN(parsed)) {
        return parsed;
      }
    }
  }
  return 1;
}

function toText(value: unknown): string {
  if (typeof value === "string") {
    return value;
  }
  if (value instanceof Buffer) {
    return value.toString("utf-8");
  }
  return "";
}
